### PR TITLE
draft: feat: fused loss and logit to logprob conversion

### PR DESCRIPTION
**Pull request by jiemingz**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Comment by euronymous-aithal** [comment]
@guyueh1  can you please review this ?

File: nemo_rl/distributed/model_utils.py [hunk 1]
```diff
@@ -18,7 +18,6 @@
 from torch.distributed.tensor import DTensor, distribute_tensor
 
 
-@torch.no_grad()
```
Comment (guyueh1) [inline nemo_rl/distributed/model_utils.py:21]: is it necessary to remove the @torch.no_grad()?

File: nemo_rl/distributed/model_utils.py [hunk 2]
```diff
@@ -175,6 +177,22 @@ def forward(  # pyrefly: ignore[bad-override]  Always ignore torch.autograd.Func
...
 
+            vocab_parallel_logits_chunk = vocab_parallel_logits[:, chunk_start:chunk_end, :]
+            target_chunk = target[:, chunk_start:chunk_end]
+
+            # Create a mask of valid vocab ids (1 means it needs to be masked).
+            target_mask_chunk = (target_chunk < vocab_start_index) | (target_chunk >= vocab_end_index)
+            masked_target_chunk = target - vocab_start_index
+            masked_target_chunk[target_mask_chunk] = 0
+
+            distributed_logprob_forward(
```
Comment (guyueh1) [inline nemo_rl/distributed/model_utils.py:188]: why is this not returning anything? and is the consequent line 196-217 necessary or should be removed?

File: nemo_rl/distributed/model_utils.py [hunk 3]
```diff
@@ -229,7 +247,7 @@ def backward(
             logits = vocab_parallel_logits[:, chunk_start:chunk_end, :]
             logits = logits.to(dtype=torch.float32)
 
-            softmax_output = _compute_distributed_log_softmax(
+            softmax_output = _distributed_logprob_forward(
```
Comment (guyueh1) [inline nemo_rl/distributed/model_utils.py:250]: this name is not found, is it `distributed_logprob_forward`?

**Review by guyueh1 — COMMENTED** [review]
Overall like the idea of doing torch compile full graph on the logprob function, but since the softmax output is still kept, the memory will still be huge, we may need another path to completely remove the overhead; is this just saving the memory for mask tensor, log_softmax tensor, can you provide data how much it is saving? 

also resolve the comments please
